source/Ship.cpp

```diff
@@ -3788,20 +3805,91 @@ const vector<weak_ptr<Ship>> &Ship::GetEscorts() const
 // cues and try to stay with it when it lands or goes into hyperspace.
 void Ship::AddEscort(Ship &ship)
 {
-	escorts.push_back(ship.shared_from_this());
+	auto ship_shared = ship.shared_from_this();
+	escorts.push_back(ship_shared);
+	TuneForEscort(ship_shared);
 }
 
 
 
 void Ship::RemoveEscort(const Ship &ship)
 {
+	// Reset the cached value if we are removing the slowest escort, we will
+	// re-scan the list of escorts to set the escorts velocity based on
+	// remaining active escorts.
+	shared_ptr<Ship> slowest = slowestEscort.lock();
+	bool findSlowest = (!slowest || (&ship == slowest.get()));
+	if(findSlowest){
+		escortsVelocity = -1.;
+		slowestEscort.reset();
+	}
+
 	auto it = escorts.begin();
-	for( ; it != escorts.end(); ++it)
-		if(it->lock().get() == &ship)
+	while(it != escorts.end())
+	{
+		auto escort = it->lock();
+		if(escort.get() == &ship)
 		{
-			escorts.erase(it);
-			return;
+			it = escorts.erase(it);
+			// If we are not removing the slowest escort, then we don't
+			// need to finish the loop to tune all other escorts.
+			if(!findSlowest)
+				return;
+		}
+		else
+		{
+			if(escort && findSlowest)
+				TuneForEscort(escort);
+			++it;
+		}
+	}
+}
+
+
+
+// (Re)Register escorts, for example because some escort got destroyed.
+void Ship::TuneForEscorts()
+{
+	// Check if we have a valid slowest escort. If we have one, then we don't
+	// need to rescan for the next slowest escort.
+	shared_ptr<Ship> slowest = slowestEscort.lock();
+	if(slowest && !slowest->IsDestroyed() && government == slowest->GetGovernment())
+		return;
+
+	// Reset this value, we will re-scan the list of escorts to set the
+	// escorts velocity based on remaining active escorts.
+	escortsVelocity = -1.;
+	slowestEscort.reset();
```

Comment:
A struct is more readable, and if we do this, then I feel it is nicer to also add the vector with the escorts themselves to the struct (since modifications to that vector should also always trigger a check for the speeds).

Reply:
Yes, that would be in somewhat in line with implementing a "ship < squadron < fleet < government" AI hierarchy I mentioned in a previous PR / discord.

Reply:
I added the struct (and I also tried a different approach by an extended vector class that tracks the cruise-speed, but such extended vector would result in some code-duplication for the government code, so I decided to go for the struct).